src/Tools/dotnet-monitor/StartupHook/StartupHookEndpointInfoSourceCallback.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.StartupHook
+{
+    internal sealed class StartupHookEndpointInfoSourceCallbacks : IEndpointInfoSourceCallbacks
+    {
+        private readonly IInProcessFeatures _inProcessFeatures;
+        private readonly StartupHookValidator _startupHookValidator;
+
+
+        public StartupHookEndpointInfoSourceCallbacks(
+            IInProcessFeatures inProcessFeatures,
+            StartupHookValidator startupHookValidator)
+        {
+            _inProcessFeatures = inProcessFeatures;
+            _startupHookValidator = startupHookValidator;
+        }
+
+        Task IEndpointInfoSourceCallbacks.OnAddedEndpointInfoAsync(IEndpointInfo endpointInfo, CancellationToken cancellationToken)
+        {
+            return Task.CompletedTask;
+        }
+
+        Task IEndpointInfoSourceCallbacks.OnBeforeResumeAsync(IEndpointInfo endpointInfo, CancellationToken cancellationToken)
+        {
+            if (_inProcessFeatures.IsStartupHookRequired)
+            {
+                _ = _startupHookValidator.ApplyStartupHook(endpointInfo, cancellationToken); // Not currently doing anything with the result
```

Comment:
I think the StartupHookValidator would be simplified if you:
1) Call `CheckAsync` without logging instructions
1) If true, immediately return
1) Call `ApplyStartupHook`
1) If true, immediately return
1) Call `CheckAsync` with logging instructions

The `CheckAsync` method shouldn't check the runtime version but unconditionally check if the environment variable exists. If it make it more clear, rename this method to `CheckEnvironmentAsync`.

The `ApplyStartupHook` method shouldn't invoke the `CheckAsync` method but merely call `ApplyStartupHook` if it's 8.0 or newer; return false on exception or if lower than 8.0

Reply:
Sure, I can separate those concerns - if I'm interpreting this correctly, then `CheckAsync` (or `CheckEnvironmentAsync`) will fail in most cases for 8+ since the env var won't be set, even though we'll have already (hopefully) applied the startup hook - is that the behavior we want?

Reply:
Yes, if the user configured the environment variable manually, that should take precedence over applying the startup hook through the diagnostic communication channel.

Reply:
I might not have explained that correctly (or maybe I'm just misunderstanding) - assuming [this call](https://github.com/dotnet/dotnet-monitor/blob/main/src/Tools/dotnet-monitor/Exceptions/ExceptionsService.cs#L70) still exists with the updated `CheckEnvironmentAsync` method, it'll fail for 8, and then `ExceptionsService` will throw a `NotSupportedException` - which means we wouldn't start the pipeline, and exceptions wouldn't work.

Reply:
Yup, you are correct. What we need is state stored per-process to track whether a startup hook was actually validated or applied. That state can be added to the `StartupHookEndpointInfoSourceCallback` class, the class elevated to a service, and the ExceptionsService would ask that service if the startup had been applied correctly.

Reply:
The latest iteration includes something along these lines - let me know if you want me to take a different approach until we come up with a more encompassing solution.